File: wilsonChan/assignments/reverseLinked/lc92/lc92.py
# Problem Statement #

# Given the head of a LinkedList and two positions ‘p’ and ‘q’, reverse the LinkedList from position ‘p’ to ‘q’.

'''

Given:
-p and q positions
-head node

Result:
-swap the nodes between p and q position



EXAMPLE # 1:
[1] [2] [3] [4] [5] [6] [7] [8] [9] [10]
         p               q
 H
        [7] [6] [5] [4] [3]
            Swapping
            p <=> q
            3 <=> 7
            4 <=> 6
            5 <=> 5

EXAMPLE # 2
(head, 2, 4)
[1] [2] [3] [4] [5]
 H   p       q
    [4] [3] [2]
      Swapping
      p <=> q
      2 <=> 4
      3 <=> 3

Misc. Notes
-If next node is null = done



Variables:
-head (starting node)
-current (what node position we're on)
-p (left node position)
-q (right node position)




[1] [2] [3] [4] [5] [6] [7] [8] [9] [10]
         p               q
 H
 C


STEPS:
i = 1    (after the head node)

-While loop to traverse until p
  -i < p    (after head node and p node)
  -i++
  -prev = null
  


-while loop p and q (swappage here)
  p = null
  next = current.next
  current.next = prev
  prev = current

  current = next

  i++

  return head


'''


from __future__ import print_function


class Node:
  def __init__(self, value, next=None):
    self.value = value
    self.next = next

  def print_list(self):
    temp = self
    while temp is not None:
      print(temp.value, end=" ")
      temp = temp.next
    print()


def reverse_sub_list(head, p, q):
  # TODO: Write your code here


# [1] [2] [3] [4] [5] [6] [7] [8] [9] [10]
#          p               q

#  H

  #Initialize Variables
  prev = None
  current = head
  pos = 1


  #while loop 1 (traverse between starting node and p)
  while pos < p and current:
    prev = current
    current = current.next
    pos += 1
  
  start = current
  end = None
  
  #while loop 2 (traverse between p and q)
  while pos <= q and current:
    next = current.next
    current.next = end
    end = current
    current = next
    pos += 1

  start.next = current

  if not prev:
    head = end
  else:
    prev.next = end

  return head





##################################################

  if head is None or p==q:
    return head

  p1_end, p2_start = head, head
  i = 0
  while p2_start and i< p-1:
    p1_end = p2_start
    p2_start = p2_start.next
    i += 1

  p2_start.print_list()

  current = p2_start
  previous, next = None, None
  i = 0
  while current and i < q-p+1:
    next = current.next
    current.next = previous
    previous = current
    current = next
    i += 1

  if p1_end:
    p1_end.next = previous
  else:
    head = previous

  p2_start.next = current

  return head


def main():
  head = Node(1)
  head.next = Node(2)
  head.next.next = Node(3)
  head.next.next.next = Node(4)
  head.next.next.next.next = Node(5)

  print("Nodes of original LinkedList are: ", end='')
  head.print_list()
  result = reverse_sub_list(head, 2, 4)
  print("Nodes of reversed LinkedList are: ", end='')
  result.print_list()

#############################################
#MY EXAMPLE
  # head = Node(1)
  # head.next = Node(2)
  # head.next.next = Node(3)
  # head.next.next.next = Node(4)
  # head.next.next.next.next = Node(5)
  # head.next.next.next.next.next = Node(6)
  # head.next.next.next.next.next.next = Node(7)

  # print("Nodes of original LinkedList are: ", end='')
  # head.print_list()
  # result = reverse_sub_list(head, 2, 6)
  # print("Nodes of reversed LinkedList are: ", end='')
  # result.print_list()
#############################################



main()



# Solution
# -----
# def reverse_sub_list(head, p, q):
#   if p == q:
#     return head

#   # after skipping 'p-1' nodes, current will point to 'p'th node
#   current, previous = head, None
#   i = 0
#   while current is not None and i < p - 1:
#     previous = current
#     current = current.next
#     i += 1

#   # we are interested in three parts of the LinkedList, the part before index 'p',
#   # the part between 'p' and 'q', and the part after index 'q'
#   last_node_of_first_part = previous
#   # after reversing the LinkedList 'current' will become the last node of the sub-list
#   last_node_of_sub_list = current
#   next = None  # will be used to temporarily store the next node

#   i = 0
#   # reverse nodes between 'p' and 'q'
#   while current is not None and i < q - p + 1:
#     next = current.next
#     current.next = previous
#     previous = current
#     current = next
#     i += 1

#   # connect with the first part
#   if last_node_of_first_part is not None:
#     # 'previous' is now the first node of the sub-list
#     last_node_of_first_part.next = previous
#   # this means p == 1 i.e., we are changing the first node (head) of the LinkedList
#   else:
#     head = previous

#   # connect with the last part
#   last_node_of_sub_list.next = current
#   return head

# -----

# Time complexity #
# The time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

# Space complexity #
# We only used constant space, therefore, the space complexity of our algorithm is O(1).

File: tiantianXia/assignments/reverseLinked/lc92/lc92.js
// Problem Statement #

// Given the head of a LinkedList and two positions ‘p’ and ‘q’, reverse the LinkedList from position ‘p’ to ‘q’.

class Node {
  constructor(value, next=null){
    this.value = value;
    this.next = next;
  }

  get_list() {
    let result = "";
    let temp = this;
    while (temp !== null) {
      result += temp.value + " ";
      temp = temp.next;
    }
    return result;
  }
};

/**
 * inputs: 
 * @param {*} head : the head of linked list
 * @param {*} p: integer for the starting of the linked list
 * @param {*} q: integer for the end of the linked list
 * 
 * output:
 * @returns head: head of the reversed linked list
 * 
 * step 1: set previous, current, next: to initiate the pointer that can point to the current node and the node before and after
 * step 2: use while loop traverse all the node
 * step 3: compare if the node value is p run the while loop and start reverse the sub linked list
 * step 4: after reverse the linked list connect the node before reverse and the node after the reverse
 * 
 * variable current, previous, next to initiate the pointer that can point to the current node and previous node
 *          nodeBeforeP: store the node before the reverse
 *          nodeP: store the node for the first sub linked list that need to point to the node after the reverse
 * 
 * a while loop traverse all the node if the node is not null keep running
 *    compare if the current node value is equal to p
 *      store the current node to the nodeP that need to use after reverse
 *      store the previous node to the nodeBeforeP that need to use after reverse and point to the node before the reversed node
 *      
 *       while the current is not null and traverse the node between p to q value
 *            store the current.next node to next
 *            store the previous node to the current.next
 *            store the current to the previous
 *            store the next to the current
 *        end loop for reverse the sub linked list
 *        
 *        link the head of reversed list to nodeBeforeP
 *        link the end of the reversed sub linked list to current since the while loop end up with the next node of q
 *    end if 
 *  
 *    store the current to the previous
 *    update the current to the current.next
 */

const reverse_sub_list = function(head, p, q) {
  // TODO: Write your code here
  let next = null;
  let current = head;
  let previous = null;
  let nodeBeforeP = null;
  let nodeP = null;
  let count = 0;

  while(current !== null){

      if(current.value === p){
        nodeP = current;
        nodeBeforeP = previous;

        while(current !== null && count < 2){
          next = current.next;
          current.next = previous;
          previous = current;
          current  = next;

          if(current.value === q || count > 0){
            count ++;
          }
        }

        nodeBeforeP.next = previous;
        nodeP.next = current;

      }

      previous = current;
      current = current.next;
  }

  return head;
};


head = new Node(1)
head.next = new Node(2)
head.next.next = new Node(3)
head.next.next.next = new Node(4)
head.next.next.next.next = new Node(5)

console.log(`Nodes of original LinkedList are: ${head.get_list()}`)// 1, 2, 3, 4, 5
console.log(`Nodes of reversed LinkedList are: ${reverse_sub_list(head, 2, 4).get_list()}`) // 1, 4, 3, 2, 5


// Solution
// -----
// function reverse_sub_list(head, p, q) {
//   if (p === q) {
//     return head;
//   }

//   // after skipping 'p-1' nodes, current will point to 'p'th node
//   let current = head,
//     previous = null;
//   let i = 0;
//   while (current !== null && i < p - 1) {
//     previous = current;
//     current = current.next;
//     i += 1;
//   }

//   // we are interested in three parts of the LinkedList, the part before index 'p',
//   // the part between 'p' and 'q', and the part after index 'q'
//   const last_node_of_first_part = previous;
//   // after reversing the LinkedList 'current' will become the last node of the sub-list
//   const last_node_of_sub_list = current;
//   let next = null; // will be used to temporarily store the next node

//   i = 0;
//   // reverse nodes between 'p' and 'q'
//   while (current !== null && i < q - p + 1) {
//     next = current.next; // 2  3
//     current.next = previous;// null 1
//     previous = current; // 1 2
//     current = next; // 2 3
//     i += 1;
//   }

//   // connect with the first part
//   if (last_node_of_first_part !== null) {
//     // 'previous' is now the first node of the sub-list
//     last_node_of_first_part.next = previous;
//     // this means p === 1 i.e., we are changing the first node (head) of the LinkedList
//   } else {
//     head = previous;
//   }

//   // connect with the last part
//   last_node_of_sub_list.next = current;
//   return head;
// }

// -----

// Time complexity #
// The time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

// Space complexity #
// We only used constant space, therefore, the space complexity of our algorithm is O(1).

/**
 * inputs:
 *  head: head of the linkedlist
 *  p: a vlue of the end of the sub linked list
 *  q: a value of the head of the sub linked list
 * 
 * out:
 *  head: the head of the linked list
 * 
 * step 1: ues while loop to travers all the node
 * setp 2: while found p, start sub while to reverse the sub linked list until found q
 * step 3: conect p to next conect q to previes
 * 
 * varible prev initial as null , current initial as head, next is head, before_p, 
 * 
 * while the current is not null traverse to the next
 *    if found the p start a while loop
 *       
 *      while current.value is not equal q run the loop
 *          let next become current.next
 *          let current.next become previous
 *          let previous become currnt
 *          let current become next
 * end loop
 * return head
 * 
 * edge case: if head is null => null
 *            
 *     
 */

//  const reverse_sub_list = function(head, p, q) {
//   // TODO: Write your code here
//   let previous = null;
//   let current = head;
//   let next = current.next;
//   let nodeBeforeP = null;
//   let nodeP = null;
//   let count = 0;

//   while(current !== null){

//     if(current.value === p){ // 2
//       nodeBeforeP = previous;
//       nodeP = current;
//       previous = null;

//       while(current !== null && count < 2 ){
//         next = current.next; //  3 4
//         current.next = previous; // 1 2
//         previous = current; // 2 3
//         current = next; // 3 4

//         if(current.value === q || count > 0){
//           count ++;
//         }
//       }
      
//       nodeBeforeP.next = previous;
//       nodeP.next = current;
//       break;

//     }else{

//     previous = current; // 1 4 5
//     current = current.next;// 2 5 null 
//     }

//   }

//   return head;
// };